analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs

```diff
@@ -18,158 +18,147 @@
+    private static IMethodSymbol GetEnumerableExtensionSymbol(InvocationExpressionSyntax invocation, SemanticModel semanticModel) =>
+        invocation.GetMethodCallIdentifier() is { } methodName
+        && CastIEnumerableMethods.Contains(methodName.ValueText)
+        && semanticModel.GetSymbolInfo(invocation).Symbol is IMethodSymbol methodSymbol
+        && methodSymbol.IsExtensionOn(KnownType.System_Collections_IEnumerable)
+            ? methodSymbol
+            : null;
 
-        private static ITypeSymbol GetGenericTypeArgument(ITypeSymbol type) =>
-            type is INamedTypeSymbol returnType && returnType.Is(KnownType.System_Collections_Generic_IEnumerable_T)
-                ? returnType.TypeArguments.Single()
-                : null;
+    private static ITypeSymbol GetGenericTypeArgument(ITypeSymbol type) =>
+        type is INamedTypeSymbol returnType && returnType.Is(KnownType.System_Collections_Generic_IEnumerable_T)
+            ? returnType.TypeArguments.Single()
+            : null;
 
-        private static bool CanHaveNullValue(ITypeSymbol type) => type.IsReferenceType || type.Name == "Nullable";
+    private static bool CanHaveNullValue(ITypeSymbol type) =>
+        type.IsReferenceType || type.Name == nameof(Nullable<int>);
 
-        private static Location GetReportLocation(InvocationExpressionSyntax invocation, bool methodCalledAsStatic)
+    private static Location GetReportLocation(InvocationExpressionSyntax invocation, bool methodCalledAsStatic) =>
+        invocation.Expression switch
         {
-            if (!(invocation.Expression is MemberAccessExpressionSyntax memberAccess))
+            MemberAccessExpressionSyntax memberAccess => methodCalledAsStatic
+                ? memberAccess.GetLocation()
+                : memberAccess.OperatorToken.CreateLocation(invocation),
+            _ => invocation.Expression.GetLocation()
+        };
+
+    private static ITypeSymbol GetElementType(InvocationExpressionSyntax invocation, IMethodSymbol methodSymbol, SemanticModel semanticModel)
+    {
+        return CollectionExpression() is { } collection
+            ? semanticModel.GetTypeInfo(collection).Type switch
             {
-                return invocation.Expression.GetLocation();
+                INamedTypeSymbol { TypeArguments: { Length: 1 } typeArguments } => typeArguments.First(),
+                IArrayTypeSymbol { Rank: 1 } arrayType => arrayType.ElementType,    // casting is necessary for multidimensional arrays
+                _ => null
             }
+            : null;
 
-            return methodCalledAsStatic
-                ? memberAccess.GetLocation()
-                : memberAccess.OperatorToken.CreateLocation(invocation);
-        }
-
-        private static ITypeSymbol GetElementType(InvocationExpressionSyntax invocation, IMethodSymbol methodSymbol,
-            SemanticModel semanticModel)
+        ExpressionSyntax CollectionExpression()
```

Comment:
Coverage is bad in this method. You probably need some non-compiling code to increase coverage here. It tests "Cast" and "TypeOf" for different kinds of invocations. You may try
```cs
col?.Cast<int>(); // MemberBinding
colOfcol[0].Cast<int>(); // ElementAccess
colOfcol[0]?.Cast<int>(); // ElementBinding
func().Cast<int>(); // func is Func<IEnumerable<int>>
func()?.Cast<int>()?.Where(x=>true);
Enumerable.Cast<int>(); // compiler error (test for FirstAndDefault in the unreduced extension method call path)
```